File: compat/suites/dotnet-sdk/Groups/SnsGroup.cs
using Amazon.SimpleNotificationService.Model;
using Amazon.SQS.Model;
using OvercastCompat.Clients;
using OvercastCompat.Harness;

namespace OvercastCompat.Groups;

public sealed class SnsGroup(AwsClients clients) : IServiceGroup
{
    public IReadOnlyDictionary<string, TestFn> Impls() => new Dictionary<string, TestFn>(StringComparer.Ordinal)
    {
        ["CreateTopic"] = CreateTopicAsync,
        ["ListTopics"] = ListTopicsAsync,
        ["GetTopicAttributes"] = GetTopicAttributesAsync,
        ["SetTopicAttributes"] = SetTopicAttributesAsync,
        ["DeleteTopic"] = DeleteTopicAsync,
        ["Publish"] = PublishAsync,
        ["PublishWithAttributes"] = PublishWithAttributesAsync,
        ["PublishBatch"] = PublishBatchAsync,
        ["SubscribeSQS"] = SubscribeSQSAsync,
        ["ListSubscriptionsByTopic"] = ListSubscriptionsByTopicAsync,
        ["GetSubscriptionAttributes"] = GetSubscriptionAttributesAsync,
        ["PublishDeliveredToSQS"] = PublishDeliveredToSQSAsync,
        ["SetSubscriptionAttributes"] = SetSubscriptionAttributesAsync,
        ["Unsubscribe"] = UnsubscribeAsync,
    };

    public IReadOnlyDictionary<string, SetupFn> Setups() => new Dictionary<string, SetupFn>(StringComparer.Ordinal)
    {
        ["sns-publish"] = SetupPublishAsync,
        ["sns-subscriptions"] = SetupSubscriptionsAsync,
    };

    public IReadOnlyDictionary<string, SetupFn> Teardowns() => new Dictionary<string, SetupFn>(StringComparer.Ordinal)
    {
        ["sns-topics"] = TeardownTopicsAsync,
        ["sns-publish"] = TeardownPublishAsync,
        ["sns-subscriptions"] = TeardownSubscriptionsAsync,
    };

    // ── sns-topics ──

    private async Task CreateTopicAsync(TestContext context)
    {
        var name = $"{context.RunId}-sns-topics";
        var response = await clients.SNS().CreateTopicAsync(new CreateTopicRequest { Name = name });
        var arn = response.TopicArn;
        Assertions.NotBlank(arn, "CreateTopic: TopicArn");
        context.Set("snsTopicArn", arn);
        var topics = await clients.SNS().ListTopicsAsync(new ListTopicsRequest());
        Assertions.True(topics.Topics.Any(t => t.TopicArn == arn), $"CreateTopic: topic {name} not found in ListTopics (runId={context.RunId})");
    }

    private async Task ListTopicsAsync(TestContext context)
    {
        var topicArn = RequireTopicArn(context, "snsTopicArn");
        var response = await clients.SNS().ListTopicsAsync(new ListTopicsRequest());
        Assertions.True(response.Topics.Any(t => t.TopicArn == topicArn), $"ListTopics: topic {topicArn} not found (runId={context.RunId})");
    }

    private async Task GetTopicAttributesAsync(TestContext context)
    {
        var topicArn = RequireTopicArn(context, "snsTopicArn");
        var response = await clients.SNS().GetTopicAttributesAsync(new GetTopicAttributesRequest { TopicArn = topicArn });
        Assertions.True(response.Attributes.Count > 0, "GetTopicAttributes: expected non-empty attributes");
        Assertions.NotBlank(response.Attributes["TopicArn"], "GetTopicAttributes: TopicArn");
    }

    private async Task SetTopicAttributesAsync(TestContext context)
    {
        var topicArn = RequireTopicArn(context, "snsTopicArn");
        await clients.SNS().SetTopicAttributesAsync(new SetTopicAttributesRequest
        {
            TopicArn = topicArn,
            AttributeName = "DisplayName",
            AttributeValue = "compat-display",
        });
        var attrs = await clients.SNS().GetTopicAttributesAsync(new GetTopicAttributesRequest { TopicArn = topicArn });
        Assertions.Equal("compat-display", attrs.Attributes["DisplayName"], "SetTopicAttributes: DisplayName");
    }

    private async Task DeleteTopicAsync(TestContext context)
    {
        var name = $"{context.RunId}-sns-del";
        var create = await clients.SNS().CreateTopicAsync(new CreateTopicRequest { Name = name });
        var arn = create.TopicArn;
        await clients.SNS().DeleteTopicAsync(new DeleteTopicRequest { TopicArn = arn });
        var topics = await clients.SNS().ListTopicsAsync(new ListTopicsRequest());
        Assertions.False(topics.Topics.Any(t => t.TopicArn == arn), $"DeleteTopic: topic {name} still present (runId={context.RunId})");
    }

    private async Task TeardownTopicsAsync(TestContext context)
    {
        var prefix = $"{context.RunId}-sns";
        await DeleteTopicByPrefixAsync(prefix);
    }

    // ── sns-publish ──

    private async Task SetupPublishAsync(TestContext context)
    {
        var name = $"{context.RunId}-sns-pub";
        var response = await clients.SNS().CreateTopicAsync(new CreateTopicRequest { Name = name });
        context.Set("snsPubTopicArn", response.TopicArn);
    }

    private async Task PublishAsync(TestContext context)
    {
        var topicArn = RequireTopicArn(context, "snsPubTopicArn");
        var response = await clients.SNS().PublishAsync(new PublishRequest
        {
            TopicArn = topicArn,
            Message = "hello from dotnet-sdk",
            Subject = "dotnet-test",
        });
        Assertions.NotBlank(response.MessageId, "Publish: MessageId");
    }

    private async Task PublishWithAttributesAsync(TestContext context)
    {
        var topicArn = RequireTopicArn(context, "snsPubTopicArn");
        var response = await clients.SNS().PublishAsync(new PublishRequest
        {
            TopicArn = topicArn,
            Message = "message with attrs",
            MessageAttributes = new Dictionary<string, Amazon.SimpleNotificationService.Model.MessageAttributeValue>
            {
                ["color"] = new Amazon.SimpleNotificationService.Model.MessageAttributeValue { DataType = "String", StringValue = "red" },
                ["size"] = new Amazon.SimpleNotificationService.Model.MessageAttributeValue { DataType = "Number", StringValue = "42" },
            },
        });
        Assertions.NotBlank(response.MessageId, "PublishWithAttributes: MessageId");
    }

    private async Task PublishBatchAsync(TestContext context)
    {
        var topicArn = RequireTopicArn(context, "snsPubTopicArn");
        var response = await clients.SNS().PublishBatchAsync(new PublishBatchRequest
        {
            TopicArn = topicArn,
            PublishBatchRequestEntries =
            [
                new PublishBatchRequestEntry { Id = "1", Message = "batch-1" },
                new PublishBatchRequestEntry { Id = "2", Message = "batch-2" },
            ],
        });
        Assertions.GreaterThanOrEqual(2, response.Successful.Count, "PublishBatch: expected >= 2 successful");
        Assertions.True(response.Failed.Count == 0, "PublishBatch: expected 0 failed");
    }

    private async Task TeardownPublishAsync(TestContext context)
    {
        var prefix = $"{context.RunId}-sns-pub";
        await DeleteTopicByPrefixAsync(prefix);
    }

    // ── sns-subscriptions ──

    private async Task SetupSubscriptionsAsync(TestContext context)
    {
        var topicName = $"{context.RunId}-sns-sub";
        var queueName = $"{context.RunId}-sns-sub-q";

        var topicResponse = await clients.SNS().CreateTopicAsync(new CreateTopicRequest { Name = topicName });
        context.Set("snsSubTopicArn", topicResponse.TopicArn);

        var queueResponse = await clients.SQS().CreateQueueAsync(new CreateQueueRequest { QueueName = queueName });
        context.Set("snsSubQueueUrl", queueResponse.QueueUrl);
    }

    private async Task SubscribeSQSAsync(TestContext context)
    {
        var topicArn = RequireTopicArn(context, "snsSubTopicArn");
        var queueUrl = RequireQueueUrl(context, "snsSubQueueUrl");

        var qAttrs = await clients.SQS().GetQueueAttributesAsync(new GetQueueAttributesRequest
        {
            QueueUrl = queueUrl,
            AttributeNames = ["QueueArn"],
        });
        var queueArn = qAttrs.Attributes["QueueArn"];

        var response = await clients.SNS().SubscribeAsync(new SubscribeRequest
        {
            TopicArn = topicArn,
            Protocol = "sqs",
            Endpoint = queueArn,
        });
        var subArn = response.SubscriptionArn;
        Assertions.NotBlank(subArn, "SubscribeSQS: SubscriptionArn");
        context.Set("snsSubArn", subArn);
    }

    private async Task ListSubscriptionsByTopicAsync(TestContext context)
    {
        var topicArn = RequireTopicArn(context, "snsSubTopicArn");
        var subArn = RequireSubArn(context);

        var response = await clients.SNS().ListSubscriptionsByTopicAsync(new ListSubscriptionsByTopicRequest
        {
            TopicArn = topicArn,
        });
        Assertions.True(response.Subscriptions.Any(s => s.SubscriptionArn == subArn), $"ListSubscriptionsByTopic: subscription {subArn} not found (runId={context.RunId})");
    }

    private async Task GetSubscriptionAttributesAsync(TestContext context)
    {
        var subArn = RequireSubArn(context);
        var response = await clients.SNS().GetSubscriptionAttributesAsync(new GetSubscriptionAttributesRequest
        {
            SubscriptionArn = subArn,
        });
        Assertions.True(response.Attributes.Count > 0, "GetSubscriptionAttributes: expected non-empty attributes");
        Assertions.NotBlank(response.Attributes["Protocol"], "GetSubscriptionAttributes: Protocol");
    }

    private async Task PublishDeliveredToSQSAsync(TestContext context)
    {
        var topicArn = RequireTopicArn(context, "snsSubTopicArn");
        var queueUrl = RequireQueueUrl(context, "snsSubQueueUrl");

        var pub = await clients.SNS().PublishAsync(new PublishRequest
        {
            TopicArn = topicArn,
            Message = "delivery-test",
        });
        Assertions.NotBlank(pub.MessageId, "PublishDeliveredToSQS: MessageId");

        List<Message> messages = new();
        for (var i = 0; i < 10; i++)
        {
            var recv = await clients.SQS().ReceiveMessageAsync(new ReceiveMessageRequest
            {
                QueueUrl = queueUrl,
                MaxNumberOfMessages = 10,
                WaitTimeSeconds = 1,
            });
            messages = recv.Messages;
            if (messages.Count > 0)
            {
                break;
            }
        }
        Assertions.True(messages.Count > 0, "PublishDeliveredToSQS: expected at least 1 message delivered via SNS->SQS");
    }

    private async Task SetSubscriptionAttributesAsync(TestContext context)
    {
        var subArn = RequireSubArn(context);
        await clients.SNS().SetSubscriptionAttributesAsync(new SetSubscriptionAttributesRequest
        {
            SubscriptionArn = subArn,
            AttributeName = "RawMessageDelivery",
            AttributeValue = "true",
        });
        var attrs = await clients.SNS().GetSubscriptionAttributesAsync(new GetSubscriptionAttributesRequest
        {
            SubscriptionArn = subArn,
        });
        Assertions.Equal("true", attrs.Attributes["RawMessageDelivery"], "SetSubscriptionAttributes: RawMessageDelivery");
    }

    private async Task UnsubscribeAsync(TestContext context)
    {
        var subArn = RequireSubArn(context);
        await clients.SNS().UnsubscribeAsync(new UnsubscribeRequest { SubscriptionArn = subArn });
        // Clear the subArn so teardown does not try to unsubscribe again
        context.Set("snsSubArn", null);
    }

    private async Task TeardownSubscriptionsAsync(TestContext context)
    {
        // Unsubscribe if still active (not already unsubscribed by the Unsubscribe test)
        var subArn = context.GetString("snsSubArn");
        if (!string.IsNullOrWhiteSpace(subArn))
        {
            try { await clients.SNS().UnsubscribeAsync(new UnsubscribeRequest { SubscriptionArn = subArn }); } catch { }
        }

        // Delete the topic
        var topicPrefix = $"{context.RunId}-sns-sub";
        await DeleteTopicByPrefixAsync(topicPrefix);

        // Delete the SQS queue
        var queueUrl = context.GetString("snsSubQueueUrl");
        if (!string.IsNullOrWhiteSpace(queueUrl))
        {
            try { await clients.SQS().PurgeQueueAsync(new PurgeQueueRequest { QueueUrl = queueUrl }); } catch { }
            try { await clients.SQS().DeleteQueueAsync(new DeleteQueueRequest { QueueUrl = queueUrl }); } catch { }
        }
    }

    // ── helpers ──

    private async Task DeleteTopicByPrefixAsync(string prefix)
    {
        var response = await clients.SNS().ListTopicsAsync(new ListTopicsRequest());
        foreach (var topic in response.Topics)
        {
            if (topic.TopicArn != null && topic.TopicArn.Contains(prefix, StringComparison.OrdinalIgnoreCase))
            {
                try { await clients.SNS().DeleteTopicAsync(new DeleteTopicRequest { TopicArn = topic.TopicArn }); } catch { }
            }
        }
    }

    private static string RequireTopicArn(TestContext context, string key)
    {
        return context.GetString(key) ?? throw new InvalidOperationException($"{key} not set");
    }

    private static string RequireQueueUrl(TestContext context, string key)
    {
        return context.GetString(key) ?? throw new InvalidOperationException($"{key} not set");
    }

    private static string RequireSubArn(TestContext context)
    {
        return context.GetString("snsSubArn") ?? throw new InvalidOperationException("snsSubArn not set");
    }
}

File: compat/suites/java-sdk/src/main/java/io/overcast/compat/groups/SnsGroup.java
package io.overcast.compat.groups;

import io.overcast.compat.clients.AwsClients;
import io.overcast.compat.harness.Assertions;
import io.overcast.compat.harness.TestContext;
import io.overcast.compat.harness.TestFn;
import software.amazon.awssdk.services.sns.SnsClient;
import software.amazon.awssdk.services.sns.model.*;
import software.amazon.awssdk.services.sqs.SqsClient;
import software.amazon.awssdk.services.sqs.model.QueueAttributeName;

import java.util.Map;

/**
 * SNS compatibility test group.
 *
 * <p>Groups: sns-topics, sns-publish, sns-subscriptions.
 */
public final class SnsGroup implements ServiceGroup {

    private final AwsClients clients;

    public SnsGroup(AwsClients clients) {
        this.clients = clients;
    }

    private SnsClient sns() { return clients.sns(); }
    private SqsClient sqs() { return clients.sqs(); }

    @Override
    public Map<String, TestFn> impls() {
        return Map.ofEntries(
                Map.entry("CreateTopic",               this::createTopic),
                Map.entry("ListTopics",                this::listTopics),
                Map.entry("GetTopicAttributes",        this::getTopicAttributes),
                Map.entry("SetTopicAttributes",        this::setTopicAttributes),
                Map.entry("DeleteTopic",               this::deleteTopic),
                Map.entry("Publish",                   this::publish),
                Map.entry("PublishWithAttributes",     this::publishWithAttributes),
                Map.entry("PublishBatch",              this::publishBatch),
                Map.entry("SubscribeSQS",              this::subscribeSqs),
                Map.entry("ListSubscriptionsByTopic",  this::listSubscriptionsByTopic),
                Map.entry("GetSubscriptionAttributes", this::getSubscriptionAttributes),
                Map.entry("PublishDeliveredToSQS",     this::publishDeliveredToSqs),
                Map.entry("SetSubscriptionAttributes", this::setSubscriptionAttributes),
                Map.entry("Unsubscribe",               this::unsubscribe)
        );
    }

    @Override
    public Map<String, TestFn> setups() {
        return Map.ofEntries(
                Map.entry("sns-topics",        this::setupTopics),
                Map.entry("sns-publish",       this::setupPublish),
                Map.entry("sns-subscriptions", this::setupSubscriptions)
        );
    }

    @Override
    public Map<String, TestFn> teardowns() {
        return Map.ofEntries(
                Map.entry("sns-topics",        ctx -> deleteTopicSilently(ctx.getString("snsTopicArn"))),
                Map.entry("sns-publish",       ctx -> deleteTopicSilently(ctx.getString("snsPubTopicArn"))),
                Map.entry("sns-subscriptions", this::teardownSubscriptions)
        );
    }

    // ── sns-topics ─────────────────────────────────────────────────────────────

    private void setupTopics(TestContext ctx) throws Exception {
        String name = ctx.runId() + "-snstopic";
        var resp = sns().createTopic(r -> r.name(name));
        ctx.set("snsTopicArn", resp.topicArn());
        ctx.set("snsTopicName", name);
    }

    private void createTopic(TestContext ctx) {
        Assertions.assertNotBlank(ctx.getString("snsTopicArn"), "snsTopicArn");
    }

    private void listTopics(TestContext ctx) throws Exception {
        String arn = ctx.getString("snsTopicArn");
        var resp = sns().listTopics();
        boolean found = resp.topics().stream().anyMatch(t -> t.topicArn().equals(arn));
        Assertions.assertTrue(found, "ListTopics: topic not found (runId=" + ctx.runId() + ")");
    }

    private void getTopicAttributes(TestContext ctx) throws Exception {
        String arn = ctx.getString("snsTopicArn");
        var resp = sns().getTopicAttributes(r -> r.topicArn(arn));
        Assertions.assertTrue(resp.hasAttributes(), "GetTopicAttributes: no attributes returned");
        Assertions.assertNotBlank(resp.attributes().get("TopicArn"), "GetTopicAttributes: TopicArn");
    }

    private void setTopicAttributes(TestContext ctx) throws Exception {
        String arn = ctx.getString("snsTopicArn");
        sns().setTopicAttributes(r -> r.topicArn(arn)
                .attributeName("DisplayName")
                .attributeValue("Overcast Test Topic"));
        var resp = sns().getTopicAttributes(r -> r.topicArn(arn));
        Assertions.assertEquals("Overcast Test Topic", resp.attributes().get("DisplayName"),
                "SetTopicAttributes: DisplayName mismatch");
    }

    private void deleteTopic(TestContext ctx) throws Exception {
        String name = ctx.runId() + "-snsdel";
        var resp = sns().createTopic(r -> r.name(name));
        sns().deleteTopic(r -> r.topicArn(resp.topicArn()));
    }

    // ── sns-publish ────────────────────────────────────────────────────────────

    private void setupPublish(TestContext ctx) throws Exception {
        String name = ctx.runId() + "-snspub";
        var resp = sns().createTopic(r -> r.name(name));
        ctx.set("snsPubTopicArn", resp.topicArn());
    }

    private void publish(TestContext ctx) throws Exception {
        String arn = ctx.getString("snsPubTopicArn");
        var resp = sns().publish(r -> r.topicArn(arn).message("Hello SNS"));
        Assertions.assertNotBlank(resp.messageId(), "Publish: messageId");
    }

    private void publishWithAttributes(TestContext ctx) throws Exception {
        String arn = ctx.getString("snsPubTopicArn");
        var resp = sns().publish(r -> r.topicArn(arn)
                .message("Attributed message")
                .messageAttributes(Map.of("attr1",
                        MessageAttributeValue.builder()
                                .dataType("String")
                                .stringValue("value1")
                                .build())));
        Assertions.assertNotBlank(resp.messageId(), "PublishWithAttributes: messageId");
    }

    private void publishBatch(TestContext ctx) throws Exception {
        String arn = ctx.getString("snsPubTopicArn");
        var resp = sns().publishBatch(r -> r.topicArn(arn).publishBatchRequestEntries(
                PublishBatchRequestEntry.builder().id("m1").message("batch-msg-1").build(),
                PublishBatchRequestEntry.builder().id("m2").message("batch-msg-2").build()
        ));
        Assertions.assertEquals(2, resp.successful().size(), "PublishBatch: expected 2 successful");
    }

    // ── sns-subscriptions ─────────────────────────────────────────────────────

    private void setupSubscriptions(TestContext ctx) throws Exception {
        String topicName = ctx.runId() + "-snssub";
        String queueName = ctx.runId() + "-snssub-q";
        var topicResp = sns().createTopic(r -> r.name(topicName));
        var queueResp = sqs().createQueue(r -> r.queueName(queueName));
        var qAttrs = sqs().getQueueAttributes(r -> r
                .queueUrl(queueResp.queueUrl())
                .attributeNames(QueueAttributeName.QUEUE_ARN));
        String queueArn = qAttrs.attributesAsStrings().get("QueueArn");
        ctx.set("snsSubTopicArn", topicResp.topicArn());
        ctx.set("snsSubQueueUrl", queueResp.queueUrl());
        ctx.set("snsSubQueueArn", queueArn);
    }

    private void teardownSubscriptions(TestContext ctx) {
        deleteTopicSilently(ctx.getString("snsSubTopicArn"));
        deleteQueueSilently(ctx.getString("snsSubQueueUrl"));
    }

    private void subscribeSqs(TestContext ctx) throws Exception {
        String topicArn  = ctx.getString("snsSubTopicArn");
        String queueArn  = ctx.getString("snsSubQueueArn");
        var resp = sns().subscribe(r -> r.topicArn(topicArn).protocol("sqs").endpoint(queueArn));
        Assertions.assertNotBlank(resp.subscriptionArn(), "SubscribeSQS: subscriptionArn");
        ctx.set("snsSubArn", resp.subscriptionArn());
    }

    private void listSubscriptionsByTopic(TestContext ctx) throws Exception {
        String topicArn = ctx.getString("snsSubTopicArn");
        var resp = sns().listSubscriptionsByTopic(r -> r.topicArn(topicArn));
        Assertions.assertNotEmpty(resp.subscriptions(), "ListSubscriptionsByTopic: no subscriptions");
    }

    private void getSubscriptionAttributes(TestContext ctx) throws Exception {
        String subArn = ctx.getString("snsSubArn");
        if (subArn == null || subArn.equals("PendingConfirmation")) return;
        var resp = sns().getSubscriptionAttributes(r -> r.subscriptionArn(subArn));
        Assertions.assertTrue(resp.hasAttributes(), "GetSubscriptionAttributes: no attributes");
    }

    private void publishDeliveredToSqs(TestContext ctx) throws Exception {
        String topicArn  = ctx.getString("snsSubTopicArn");
        String queueUrl  = ctx.getString("snsSubQueueUrl");
        sns().publish(r -> r.topicArn(topicArn).message("deliver-test"));
        // Poll SQS briefly (no sleep — just try once; delivery is eventual).
        var recv = sqs().receiveMessage(r -> r.queueUrl(queueUrl)
                .maxNumberOfMessages(1).waitTimeSeconds(2));
        Assertions.assertNotEmpty(recv.messages(), "PublishDeliveredToSQS: no messages in queue after publish");
    }

    private void setSubscriptionAttributes(TestContext ctx) throws Exception {
        String subArn = ctx.getString("snsSubArn");
        if (subArn == null || subArn.equals("PendingConfirmation")) return;
        sns().setSubscriptionAttributes(r -> r.subscriptionArn(subArn)
                .attributeName("RawMessageDelivery")
                .attributeValue("true"));
    }

    private void unsubscribe(TestContext ctx) throws Exception {
        String subArn = ctx.getString("snsSubArn");
        if (subArn == null || subArn.equals("PendingConfirmation")) return;
        sns().unsubscribe(r -> r.subscriptionArn(subArn));
    }

    // ── Helpers ───────────────────────────────────────────────────────────────

    private void deleteTopicSilently(String arn) {
        if (arn == null) return;
        try { sns().deleteTopic(r -> r.topicArn(arn)); } catch (Exception ignored) {}
    }

    private void deleteQueueSilently(String url) {
        if (url == null) return;
        try { sqs().deleteQueue(r -> r.queueUrl(url)); } catch (Exception ignored) {}
    }
}
